hail/scripts/benchmark_in_batch.py

```diff
@@ -0,0 +1,293 @@
+    combine(args.output, args.files)
+
+
+def make_combine_job(b: Batch, context: str, files: List[Resource]) -> Resource:
+    j = b.new_python_job(f'combine_output_{context}')
+    j.call(combine, j.ofile, files)
+    j.always_run()
+    return j.ofile
+
+
+def combine_results_as_tree(b: Batch, branch_factor: int, results: List[Resource]) -> Resource:
+    phase_i = 1
+
+    while len(results) > 1:
+        job_id = 0
+        results = [
+            (make_combine_job(b, f'phase{phase_i}_job{job_id}', files) if len(files) > 1 else files[0])
+            for files in chunk(branch_factor, results)
+            if (job_id := job_id + 1)
+        ]
+        phase_i += 1
+
+    return results[0]
+
+
+def run_submit(args: Namespace) -> None:
+    timestamp = time.strftime('%Y-%m-%d')
+    labelled_sha = args.sha + (f'-{args.label}' if args.label is not None else '')
+    object_prefix = getattr(args, 'object-prefix')
+    output_file = P.join(object_prefix, f'{timestamp}-{labelled_sha}.jsonl')
+
+    all_benchmarks = [
+        (path, name, trial)
+        for path, name, num_jobs in list_benchmarks(args.include, args.exclude)
+        for trial in range(num_jobs)
+    ]
+
+    assert len(all_benchmarks) > 0
+
+    image = build_and_push_benchmark_image(args.image, args.artifact_uri, labelled_sha)
+
+    b = Batch(
+        name=f'benchmark-{labelled_sha}',
+        backend=hb.ServiceBackend(),
+        default_image=image,
+        default_python_image=image,
+        default_cpu='2',
+        default_storage='30G',
+        attributes={
+            'output_file': output_file,
+            'sha': args.sha,
+            'image': image,
+        },
+    )
+
+    test_permissions = make_test_storage_permissions_job(b, object_prefix, labelled_sha)
+
+    print(f'generating {len(all_benchmarks)} individual benchmark tasks')
+
+    envvars = {
+        'MKL_NUM_THREADS': '1',
+        'OPENBLAS_NUM_THREADS': '1',
+        'OMP_NUM_THREADS': '1',
+        'VECLIB_MAXIMUM_THREADS': '1',
+        'HAIL_DEV_LOWER': '1' if args.lower else None,
+        'HAIL_DEV_LOWER_ONLY': '1' if args.lower_only else None,
+        'PYSPARK_SUBMIT_ARGS': '--driver-memory 6G pyspark-shell',
+        'TMPDIR': '/io/tmp',
+    }
+
+    random.shuffle(all_benchmarks)
+    result = combine_results_as_tree(
+        b,
+        args.branch_factor,
+        [
+            make_benchmark_trial(b, envvars, path, name, trial, deps=[test_permissions])
+            for path, name, trial in all_benchmarks
+        ],
+    )
+
+    print(f'writing output to {output_file}')
+    b.write_output(result, output_file)
+
+    if args.wait:
+        b.run()
+    else:
+        b = b.run(wait=False)
+        print(f"Submitted  batch '{b.id}'.")
+
+
+def nonempty(s: str):
+    if not s:
+        raise ValueError('must be non-empty')
+    else:
+        return s
+
+
+if __name__ == '__main__':
+    parser = ArgumentParser()
+    subparser = parser.add_subparsers(title='commands')
```

Comment:
I don't think this is in scope for this PR as this looks quite complete, but I wonder what your thoughts are on `typer`. We use it for the `hailctl` CLI and I think it helps with some of the `argparse` boilerplate and error handling w.r.t. argument types

Reply:
I have none. I've just used argparse in the past but it is fairly slow. typer might be a nice change in the future to make everything consistent.